c10/core/DispatchKeySet.h

```diff
@@ -120,6 +120,65 @@ class DispatchKeySet final {
 private:
   DispatchKeySet(uint64_t repr) : repr_(repr) {}
   uint64_t repr_ = 0;
+
+public:
+  // STL iterator for DispatchKeySet. Iterates through all DispatchKeys in the
+  // set. The iterator is only invalidated by the destruction ofthe underlying
+  // DispatchKeySet as the iterator stores a pointer to the raw represenation of
+  // the DispatchKeySet.
+  class iterator {
+   public:
+    using self_type = iterator;
+    using iterator_category = std::input_iterator_tag;
+    using value_type = DispatchKey;
+    using difference_type = ptrdiff_t;
+
+    iterator(const uint64_t *data_ptr, uint8_t i=0) : data_ptr_(data_ptr), i_(i) {
+      // Go to the first key in the set
+      ++(*this);
+    }
+
+    self_type& operator++() {
+      // Create a masked version of the set representation to ignore previous
+      // keys that we've iterated through.
+      uint64_t masked_data = llvm::maskTrailingZeros<uint64_t>(i_) & *data_ptr_;
+      uint64_t firstKeyIndex = llvm::findFirstSet(masked_data);
+
+      // If there are no keys, set to end iterator value
+      if (firstKeyIndex == std::numeric_limits<uint64_t>::max() ||
+          i_ >= static_cast<uint8_t>(DispatchKey::NumDispatchKeys)) {
```

Comment:
`i_ >= static_cast<uint8_t>(DispatchKey::NumDispatchKeys)` is probably assert worthy behavior